.github/workflows/test.yml

```diff
@@ -121,26 +78,60 @@ jobs:
     - name: Restore deps cache
       uses: actions/cache@v2
       with:
-        path: deps
-        key: ${{ runner.os }}-mix-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
+        path: |
+          deps
+          _build
+        key: ${{ runner.os }}-deps-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}-git-${{ github.sha }}
         restore-keys: |
-          ${{ runner.os }}-mix-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
-          ${{ runner.os }}-mix-${{ matrix.otp }}-${{ matrix.elixir }}
+          ${{ runner.os }}-deps-${{ matrix.otp }}-${{ matrix.elixir }}-${{ hashFiles('**/mix.lock') }}
+          ${{ runner.os }}-deps-${{ matrix.otp }}-${{ matrix.elixir }}
```

Comment:
>he eventual net effect of this change will be that the workflow will take longer
Actually. no.

`actions/cache` works as described below:
1. restore-keys` is not set - only if there is a cache item with exactly the same `key`, if will be downloaded. If not, dependencies will be installed without using a cache and then new item cache will be saved.
2. `restore-keys` is set -  if there is a cache item with exactly the same `key`, this is just the same as the previous item. But if not, then for each of `restore-keys`, the latest cache item with this prefix will be used, if any. So if you have any commit with the same `mix.lock` file or even with the same Elixir+OTP versions, this cache will be used.
Actually, this is the reason why dependency install step takes only 1 sec, they are fetched from cache:
![image](https://user-images.githubusercontent.com/4661021/113686025-82b18200-96cf-11eb-8c36-1e2d39a6462d.png)

So adding one more part to `key` item will not affect cache because `action/cache` will fallback to previous cache items.

Even a fact that this part is a commit hash which is different every time will not affect fetching cached files. This will affect only saving cache which takes less than a second.

> How long is the "increase of workflow run time"?
Currently you're not using instruments like dialyzes which are messing up with `deps`/`_build` dirs content, so what's why you're not facing the issue I've described above.

But eventually you can add a step with running this took to prevent issues like https://github.com/duffelhq/paginator/pull/114.
In this case every workflow run could lead to generating files used by dialyzer, and it could take at least few minutes (in one of my projects it takes about 20 minutes). And without changing mix.lock content (for example, you've changed mix.exs or .dialyzer_ignore file), this step will be executed on every workflow run because the results are not cached.

So I propose to make this change now to prevent this issue at all.

If you don't want to use dialyzer, caching will work just like now, nothing will be changed.

Reply:
I understand how the multiple passes of [`actions/cache`](http://github.com/actions/cache) work, but many thanks for articulating that for other readers. The concern I have is that GitHub only maintain a cache for 5GB of total size, which is progressively evacuated based on age. So the method you've proposed would work but only until those older caches eventually get pushed out by the newer ones. At that point no cache hit would occur.

What you proposed makes sense when a long running task like `dialyzer` is being used. As none are currently being used in this project, I suggest waiting to make this change until a time when they are.

Reply:
> So the method you've proposed would work but only until those older caches eventually get pushed out by the newer ones. At that point no cache hit would occur.
Every commit will update cache and save it with a new key, which will be user by next run. So this cannot happen.

Reply:
> As none are currently being used in this project, I suggest waiting to make this change until a time when they are.

OK, I've removed this change.